network-mux/test/Test/Mux.hs

```diff
@@ -1219,9 +1222,62 @@ dummyAppToChannel DummyApp {daAction, daRunTime} = \_ -> do
          DummyAppSucceed -> return ((), Nothing)
          DummyAppFail    -> throwM $ MuxError (MuxShutdown Nothing) "App Fail"
 
+data DummyRestartingApps =
+    DummyRestartingResponderApps [(DummyApp, Int)]
+  | DummyRestartingInitiatorApps [(DummyApp, Int)]
+  | DummyRestartingInitiatorResponderApps [(DummyApp, Int)]
+  deriving Show
+
+instance Arbitrary DummyRestartingApps where
+    arbitrary = do
+        nums <- listOf1 $ arbitrary
+        apps <- mapM genApp $ nub nums
+        mode <- arbitrary
+        case mode of
+             InitiatorMode          -> return $ DummyRestartingInitiatorApps apps
+             ResponderMode          -> return $ DummyRestartingResponderApps apps
+             InitiatorResponderMode -> return $ DummyRestartingInitiatorResponderApps apps
+      where
+        genApp num = do
+            app <- DummyApp num DummyAppSucceed <$> arbitrary <*> arbitrary
+            restarts <- choose (0, 5)
+            return (app, restarts)
+
+
+dummyRestartingAppToChannel :: forall a m.
+                     ( MonadAsync m
+                     , MonadCatch m
+                     , MonadTimer m
+                     )
+                  => (DummyApp, a)
+                  -> (Channel m -> m ((DummyApp, a), Maybe BL.ByteString))
+dummyRestartingAppToChannel (app, r) = \_ -> do
+    threadDelay $ daRunTime app
```

Comment:
We're not using the channel, shouldn't we run some simple protocol, e.g. `req-response`.

I think It would be interesting to see that if there are left overs bytes the mux will well behave when restarting a mini-protocol.

Reply:
Since we never read any data there is always a left over byte in the channel.
Anyway a simple req-rsp protocol would be a more realistic test.

Reply:
There's `Test.Mux.ReqResp` which does not even rely on `typed-protocols`.

Reply:
The case with left over bytes in the queues is covered by the testcases `prop_mux_snd_recv` and `prop_mux_snd_recv_bi` using the reqresp protocol. Complicating these tests by adding a reqresp protocol would not increase test coverage.

Reply:
What if the left over bytes are only mistreated after restarting a protocol?

Reply:
`prop_mux_snd_recv` and `prop_mux_snd_recv_bi` restarts the protocol a random number of times.